src/core/thread/mesh_forwarder.cpp

```diff
@@ -296,6 +296,9 @@ void MeshForwarder::ScheduleTransmissionTask()
         ExitNow();
     }
 
+    // Reset offset for this message
+    mMessageNextOffset = 0;
```

Comment:
`mMessageNextOffset` should always be set as part of the call to `MeshForwarder::HandleFrameRequest()`.  In the case of an IEEE 802.15.4 Data Request message, `mMessageNextOffset` is set within `MeshForwarder::SendPoll()`.  How does initializing `mMessageNextOffset` here resolve the issue that you're seeing?